### PR TITLE
Puts tracing into a class.

### PR DESCRIPTION
**Pull request by jkff**
* Pulls out a base class from `DebugVar`, in order to have another subclass FnDebugVar that has getters and setters for lambdas. This is handy for having them call someone's methods - in this case, setter methods on the `TraceControl` class
* Centralizes tracing in the `TraceControl` class, including flushing when a new trace is started, etc. (some of that code was scattered around)
* We don't need the variable for counting trace samples as it equals trace buffer size / num variables.

The next step will be to also pull control of which variables are traced into methods of `TraceControl`, and add unit tests for tracing.<git-pr-chain>

#### Commits in this PR
1. Puts tracing into a class.

#### [PR chain](https://github.com/jlebar/git-pr-chain)
1. 👉 #385 Puts tracing into a class. 👈 **YOU ARE HERE**

</git-pr-chain>

### CODE REVIEW

File: controller/lib/debug/trace.h [hunk 1]
```diff
@@ -22,35 +22,61 @@ limitations under the License.
...
+  }
 
-// The trace period gives the period of the trace data capture
-// in units of loop cycles.
-extern int32_t tracePeriod;
+  void Flush() {
+    flags_ = 0;
+    while (traceBuffer.Get() != std::nullopt) {
+    }
+  }
```
Comment (sglow) [inline controller/lib/debug/trace.h:58]: Seems like an awfully inefficient way to flush the buffer.  That could loop 16 thousand times.  Why not just call traceBuffer.Flush()?

File: controller/lib/debug/trace.cpp [hunk 2]
```diff
@@ -32,27 +32,26 @@ limitations under the License.
 #include <stdint.h>
 
 CircBuff<uint32_t, 0x4000> traceBuffer;
```
Comment (sglow) [inline controller/lib/debug/trace.cpp:34]: May as well rename trace_buffer for consistency.  Same with traceVar array

File: controller/lib/debug/trace.h (hunk 1 above)
Comment (jkff) [reply controller/lib/debug/trace.h]: Yeah, this was copied from [here](https://github.com/RespiraWorks/VentilatorSoftware/blob/1875bcc63403ff2a111d3de8f803f42ca84594b3/controller/lib/debug/trace_cmd.cpp#L44). Fixed.

File: controller/lib/debug/trace.cpp (hunk 2 above)
Comment (jkff) [reply controller/lib/debug/trace.cpp]: Done

**Pull request merged**